pkg/operator/controller/gatewayapi/controller.go

```diff
@@ -0,0 +1,136 @@
+package gatewayapi
+
+import (
+	"context"
+	"sync"
+
+	logf "github.com/openshift/cluster-ingress-operator/pkg/log"
+
+	"k8s.io/client-go/tools/record"
+
+	configv1 "github.com/openshift/api/config/v1"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const controllerName = "gatewayapi_controller"
+
+var log = logf.Logger.WithName(controllerName)
+
+// New creates and returns a controller that creates Gateway API CRDs when the
+// appropriate featuregate is enabled.
+func New(mgr manager.Manager, config Config) (controller.Controller, error) {
+	reconciler := &reconciler{
+		client: mgr.GetClient(),
+		cache:  mgr.GetCache(),
+		config: config,
+	}
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+	if err := c.Watch(&source.Kind{Type: &configv1.FeatureGate{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// Config holds all the configuration that must be provided when creating the
+// controller.
+type Config struct {
+	// DependentControllers is a list of controllers that watch Gateway API
+	// resources.  The gatewayapi controller starts these controllers once
+	// the Gateway API CRDs have been created.
+	DependentControllers []controller.Controller
+}
+
+// reconciler reconciles gatewayclasses.
+type reconciler struct {
+	config Config
+
+	client           client.Client
+	cache            cache.Cache
+	recorder         record.EventRecorder
+	startControllers sync.Once
+}
+
+// Reconcile expects request to refer to a FeatureGate and creates or
+// reconciles the Gateway API CRDs.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log.Info("reconciling", "request", request)
+
+	var fg configv1.FeatureGate
+	if err := r.cache.Get(ctx, request.NamespacedName, &fg); err != nil {
```

Comment:
Will we allow the feature gate to be enabled and disabled?  If not, we should be able to check only once, store the value, and refer to that variable from lots of places.

Reply:
The current logic doesn't undo anything when the feature gate is turned off.  Where would it be useful to be able to refer to a stored value?